File: coding/striver_sheet/recursion/subsets.py
# Problem - https://leetcode.com/problems/subsets/

# Approach 1 - Recursive
class Solution:
    def findSubsets(self, nums, l, r, subset, result):
        if l == r:
            result.append(subset)
            return

        self.findSubsets(nums, l+1, r, [nums[l]] + subset, result)
        self.findSubsets(nums, l+1, r, subset, result)

    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        self.findSubsets(nums, 0, len(nums), [], result)
        return result                                                                        
# Time complexity - O(n * 2^n)
# Space comlexity - O(n * 2^n)
# Auxiliary space - O(n * 2^n)

File: coding/topics/recursion/subsets.cpp
// Leetcode problem - 

class Solution {
public:

    vector<vector<int>> allSubsets;

    void listAllSubsets(vector<int> inputNums, vector<int> outputNums) {
        if (inputNums.size() == 0) {
            allSubsets.push_back(outputNums);
            return;
        }

        int firstNum = inputNums[0];
        inputNums.erase(inputNums.begin());
        
        listAllSubsets(inputNums, outputNums);

        outputNums.push_back(firstNum);
        listAllSubsets(inputNums, outputNums);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        listAllSubsets(nums, vector<int> {});
        return allSubsets;
    }
};

// Time Complexity = O(2^N)
// Space Complexity = O(2^N)
